diff --git a/services/webhook/discord.go b/services/webhook/discord.go
index 7259c4a995..1f2ebf985b 100644
--- a/services/webhook/discord.go
+++ b/services/webhook/discord.go
@@ -11,11 +11,10 @@ import (
 	"net/http"
 	"net/url"
 	"regexp"
 	"strconv"
 	"strings"
-	"unicode/utf8"
 
 	webhook_model "forgejo.org/models/webhook"
 	"forgejo.org/modules/base"
 	"forgejo.org/modules/git"
 	"forgejo.org/modules/json"
@@ -217,14 +216,10 @@ func (d discordConvertor) Push(p *api.PushPayload) (DiscordPayload, error) {
 		message := strings.TrimRight(strings.SplitN(commit.Message, "\n", 1)[0], "\r")
 
 		// Escaping markdown character
 		message = escapeMarkdown(message)
 
-		// a limit of 50 is set because GitHub does the same
-		if utf8.RuneCountInString(message) > 50 {
-			message = fmt.Sprintf("%.47s...", message)
-		}
 		text += fmt.Sprintf("[%s](%s) %s - %s", commit.ID[:7], commit.URL, message, commit.Author.Name)
 		// add linebreak to each commit but the last
 		if i < len(p.Commits)-1 {
 			text += "\n"
 		}
diff --git a/services/webhook/discord_test.go b/services/webhook/discord_test.go
index b04be30bc6..9c31cda988 100644
--- a/services/webhook/discord_test.go
+++ b/services/webhook/discord_test.go
@@ -85,11 +85,11 @@ func TestDiscordPayload(t *testing.T) {
 		pl, err := dc.Push(p)
 		require.NoError(t, err)
 
 		assert.Len(t, pl.Embeds, 1)
 		assert.Equal(t, "[test/repo:test] 2 new commits", pl.Embeds[0].Title)
-		assert.Equal(t, "[2020558](http://localhost:3000/test/repo/commit/2020558fe2e34debb818a514715839cabd25e778) This is a commit summary ⚠️⚠️⚠️⚠️ containing 你好... - user1\n[2020558](http://localhost:3000/test/repo/commit/2020558fe2e34debb818a514715839cabd25e778) This is a commit summary ⚠️⚠️⚠️⚠️ containing 你好... - user1", pl.Embeds[0].Description)
+		assert.Equal(t, "[2020558](http://localhost:3000/test/repo/commit/2020558fe2e34debb818a514715839cabd25e778) This is a commit summary ⚠️⚠️⚠️⚠️ containing 你好 ⚠️⚠️️\n\nThis is the message body. - user1\n[2020558](http://localhost:3000/test/repo/commit/2020558fe2e34debb818a514715839cabd25e778) This is a commit summary ⚠️⚠️⚠️⚠️ containing 你好 ⚠️⚠️️\n\nThis is the message body. - user1", pl.Embeds[0].Description)
 		assert.Equal(t, p.Sender.UserName, pl.Embeds[0].Author.Name)
 		assert.Equal(t, setting.AppURL+p.Sender.UserName, pl.Embeds[0].Author.URL)
 		assert.Equal(t, p.Sender.AvatarURL, pl.Embeds[0].Author.IconURL)
 	})
 
