
--- a/coregrind/m_dispatch/dispatch-ppc64be-linux.S	2026-04-15 20:59:02.178021097 -0400
+++ b/coregrind/m_dispatch/dispatch-ppc64be-linux.S	2026-04-15 21:00:05.853890039 -0400
@@ -86,6 +86,10 @@
 .section ".text"
 .align   2
 .globl   VG_(disp_run_translations)
+#if _CALL_ELF == 2
+.type VG_(disp_run_translations),@function
+VG_(disp_run_translations):
+#else
 .section ".opd","aw"
 .align   3
 VG_(disp_run_translations):
@@ -94,6 +98,7 @@
 .type    .VG_(disp_run_translations),@function
 .globl   .VG_(disp_run_translations)
 .VG_(disp_run_translations):
+#endif
 	/* r3 holds two_words */
 	/* r4 holds guest_state */
         /* r5 holds host_addr */
@@ -408,6 +413,10 @@
         .section ".text"
         .align   2
         .globl   VG_(disp_cp_chain_me_to_slowEP)
+#if _CALL_ELF == 2
+        .type VG_(disp_cp_chain_me_to_slowEP),@function
+VG_(disp_cp_chain_me_to_slowEP):
+#else
         .section ".opd","aw"
         .align   3
 VG_(disp_cp_chain_me_to_slowEP):
@@ -416,6 +425,7 @@
         .type    .VG_(disp_cp_chain_me_to_slowEP),@function
         .globl   .VG_(disp_cp_chain_me_to_slowEP)
 .VG_(disp_cp_chain_me_to_slowEP):
+#endif
         /* We got called.  The return address indicates
            where the patching needs to happen.  Collect
            the return address and, exit back to C land,
@@ -433,6 +443,10 @@
         .section ".text"
         .align   2
         .globl   VG_(disp_cp_chain_me_to_fastEP)
+#if _CALL_ELF == 2
+        .type VG_(disp_cp_chain_me_to_fastEP),@function
+VG_(disp_cp_chain_me_to_fastEP):
+#else
         .section ".opd","aw"
         .align   3
 VG_(disp_cp_chain_me_to_fastEP):
@@ -441,6 +455,7 @@
         .type    .VG_(disp_cp_chain_me_to_fastEP),@function
         .globl   .VG_(disp_cp_chain_me_to_fastEP)
 .VG_(disp_cp_chain_me_to_fastEP):
+#endif
         /* We got called.  The return address indicates
            where the patching needs to happen.  Collect
            the return address and, exit back to C land,
@@ -458,6 +473,10 @@
         .section ".text"
         .align   2
         .globl   VG_(disp_cp_xindir)
+#if _CALL_ELF == 2
+        .type VG_(disp_cp_xindir),@function
+VG_(disp_cp_xindir):
+#else
         .section ".opd","aw"
         .align   3
 VG_(disp_cp_xindir):
@@ -466,6 +485,7 @@
         .type    .VG_(disp_cp_xindir),@function
         .globl   .VG_(disp_cp_xindir)
 .VG_(disp_cp_xindir):
+#endif
         /* Where are we going? */
         ld    20, OFFSET_ppc64_CIA(31)
 
@@ -589,6 +609,10 @@
 .section ".text"
         .align   2
         .globl   VG_(disp_cp_xassisted)
+#if _CALL_ELF == 2
+        .type VG_(disp_cp_xassisted),@function
+VG_(disp_cp_xassisted):
+#else
         .section ".opd","aw"
         .align   3
 VG_(disp_cp_xassisted):
@@ -597,6 +621,7 @@
         .type    .VG_(disp_cp_xassisted),@function
         .globl   .VG_(disp_cp_xassisted)
 .VG_(disp_cp_xassisted):
+#endif
         /* r31 contains the TRC */
         mr      6,31
         li      7,0
@@ -606,6 +631,10 @@
         .section ".text"
         .align   2
         .globl   VG_(disp_cp_evcheck_fail)
+#if _CALL_ELF == 2
+        .type VG_(disp_cp_evcheck_fail),@function
+VG_(disp_cp_evcheck_fail):
+#else
         .section ".opd","aw"
         .align   3
 VG_(disp_cp_evcheck_fail):
@@ -614,12 +643,17 @@
         .type    .VG_(disp_cp_evcheck_fail),@function
         .globl   .VG_(disp_cp_evcheck_fail)
 .VG_(disp_cp_evcheck_fail):
+#endif
         li      6,VG_TRC_INNER_COUNTERZERO
         li      7,0
         b       .postamble
 
         
+#if _CALL_ELF == 2
+.size VG_(disp_run_translations),.-VG_(disp_run_translations)
+#else
 .size .VG_(disp_run_translations), .-.VG_(disp_run_translations)
+#endif
 
 #endif // defined(VGP_ppc64be_linux)
 

--- a/coregrind/m_libcsetjmp.c	2026-04-15 20:59:02.185339523 -0400
+++ b/coregrind/m_libcsetjmp.c	2026-04-16 02:54:30.324374528 -0400
@@ -158,6 +158,10 @@
 ".align 2"                          "\n"
 ".p2align 4,,15"                    "\n"
 ".globl VG_MINIMAL_SETJMP"          "\n"
+#if _CALL_ELF == 2
+".type VG_MINIMAL_SETJMP, @function"   "\n"
+"VG_MINIMAL_SETJMP:"                "\n"
+#else
 ".section \".opd\",\"aw\""          "\n"
 ".align 3"                          "\n"
 "VG_MINIMAL_SETJMP:"                "\n"
@@ -166,6 +170,7 @@
 
 ".type VG_MINIMAL_SETJMP, @function"   "\n"
 ".L.VG_MINIMAL_SETJMP:"   "\n"
+#endif
 "        std     0, 0(3)"  "\n"
 "        std     1, 8(3)"  "\n"
 "        std     2, 16(3)"  "\n"
@@ -209,7 +214,10 @@
 
 
 ".globl VG_MINIMAL_LONGJMP"         "\n"
-
+#if _CALL_ELF == 2
+".type   VG_MINIMAL_LONGJMP, @function"    "\n"
+"VG_MINIMAL_LONGJMP:"               "\n"
+#else
 ".section \".opd\",\"aw\""          "\n"
 ".align 3"                          "\n"
 "VG_MINIMAL_LONGJMP:"               "\n"
@@ -218,6 +226,7 @@
 
 ".type   VG_MINIMAL_LONGJMP, @function"    "\n"
 ".L.VG_MINIMAL_LONGJMP:"            "\n"
+#endif
          // do r4 = 1
          // and park it in the restore slot for r3 (the ret reg)
 "        li      4, 1"  "\n"

--- a/coregrind/m_main.c	2026-04-15 20:59:02.185362395 -0400
+++ b/coregrind/m_main.c	2026-04-16 02:54:11.875835980 -0400
@@ -2829,6 +2829,40 @@
     ".previous\n"
 );
 #elif defined(VGP_ppc64be_linux)
+#if _CALL_ELF == 2
+asm("\n"
+    "\t.align 2\n"
+    "\t.global _start\n"
+    "\t.type _start,@function\n"
+    "_start:\n"
+    /* set up the new stack in r16 */
+    "\tlis  16,   vgPlain_interim_stack@highest\n"
+    "\tori  16,16,vgPlain_interim_stack@higher\n"
+    "\tsldi 16,16,32\n"
+    "\toris 16,16,vgPlain_interim_stack@h\n"
+    "\tori  16,16,vgPlain_interim_stack@l\n"
+    "\txor  17,17,17\n"
+    "\tlis    17,("VG_STRINGIFY(VG_STACK_GUARD_SZB)" >> 16)\n"
+    "\tori 17,17,("VG_STRINGIFY(VG_STACK_GUARD_SZB)" & 0xFFFF)\n"
+    "\txor 18,18,18\n"
+    "\tlis    18,("VG_STRINGIFY(VG_DEFAULT_STACK_ACTIVE_SZB)" >> 16)\n"
+    "\tori 18,18,("VG_STRINGIFY(VG_DEFAULT_STACK_ACTIVE_SZB)" & 0xFFFF)\n"
+    "\tadd 16,17,16\n"
+    "\tadd 16,18,16\n"
+    "\trldicr 16,16,0,59\n"
+    "\tmr 3,1\n"
+    "\tmr 1,16\n"
+    "\tlis  14,   _start_in_C_linux@highest\n"
+    "\tori  14,14,_start_in_C_linux@higher\n"
+    "\tsldi 14,14,32\n"
+    "\toris 14,14,_start_in_C_linux@h\n"
+    "\tori  14,14,_start_in_C_linux@l\n"
+    "\tmtctr 14\n"
+    "\tbctrl\n"
+    "\tnop\n"
+    "\ttrap\n"
+);
+#else
 asm("\n"
     /* PPC64 ELF ABI says '_start' points to a function descriptor.
        So we must have one, and that is what goes into the .opd section. */
@@ -2857,10 +2891,6 @@
     "\tadd 16,17,16\n"
     "\tadd 16,18,16\n"
     "\trldicr 16,16,0,59\n"
-    /* now r16 = &vgPlain_interim_stack + VG_STACK_GUARD_SZB +
-       VG_DEFAULT_STACK_ACTIVE_SZB rounded down to the nearest 16-byte
-       boundary.  And r1 is the original SP.  Set the SP to r16 and
-       call _start_in_C_linux, passing it the initial SP. */
     "\tmr 3,1\n"
     "\tmr 1,16\n"
     "\tlis  14,   _start_in_C_linux@highest\n"
@@ -2874,6 +2904,7 @@
     "\tnop\n"
     "\ttrap\n"
 );
+#endif
 #elif defined(VGP_ppc64le_linux)
 /* Little Endian uses ELF version 2 but in the future may also
  * support other ELF versions.

--- a/coregrind/m_signals.c	2026-04-15 20:59:02.184808289 -0400
+++ b/coregrind/m_signals.c	2026-04-15 21:01:00.019460468 -0400
@@ -987,6 +987,15 @@
    ".previous\n"
 
 #elif defined(VGP_ppc64be_linux)
+#if _CALL_ELF == 2
+#  define _MY_SIGRETURN(name) \
+   ".align   2\n" \
+   ".globl   my_sigreturn\n" \
+   ".type    my_sigreturn,@function\n" \
+   "my_sigreturn:\n" \
+   "	li	0, " #name "\n" \
+   "	sc\n"
+#else
 #  define _MY_SIGRETURN(name) \
    ".align   2\n" \
    ".globl   my_sigreturn\n" \
@@ -1000,6 +1009,7 @@
    ".my_sigreturn:\n" \
    "	li	0, " #name "\n" \
    "	sc\n"
+#endif
 
 #elif defined(VGP_ppc64le_linux)
 /* Little Endian supports ELF version 2.  In the future, it may

--- a/coregrind/m_syscall.c	2026-04-15 20:59:02.184886473 -0400
+++ b/coregrind/m_syscall.c	2026-04-16 02:52:53.265541190 -0400
@@ -568,6 +568,10 @@
 asm(
 ".align   2\n"
 ".globl   do_syscall_WRK\n"
+#if _CALL_ELF == 2
+".type    do_syscall_WRK,@function\n"
+"do_syscall_WRK:\n"
+#else
 ".section \".opd\",\"aw\"\n"
 ".align   3\n"
 "do_syscall_WRK:\n"
@@ -576,6 +580,7 @@
 ".type    .do_syscall_WRK,@function\n"
 ".globl   .do_syscall_WRK\n"
 ".do_syscall_WRK:\n"
+#endif
 "        std  3,-16(1)\n"  /* stash arg */
 "        ld   8, 48(3)\n"  /* sc arg 6 */
 "        ld   7, 40(3)\n"  /* sc arg 5 */

--- a/coregrind/m_syswrap/syscall-ppc64be-linux.S	2026-04-15 20:59:02.181736368 -0400
+++ b/coregrind/m_syswrap/syscall-ppc64be-linux.S	2026-04-15 21:00:15.817178906 -0400
@@ -74,14 +74,19 @@
 
 .align 2
 .globl ML_(do_syscall_for_client_WRK)
+#if _CALL_ELF == 2
+.type ML_(do_syscall_for_client_WRK),@function
+ML_(do_syscall_for_client_WRK):
+#else
 .section ".opd","aw"
 .align 3
-ML_(do_syscall_for_client_WRK):	
+ML_(do_syscall_for_client_WRK):
 .quad .ML_(do_syscall_for_client_WRK),.TOC.@tocbase,0
 .previous
 .type .ML_(do_syscall_for_client_WRK),@function
 .globl .ML_(do_syscall_for_client_WRK)
 .ML_(do_syscall_for_client_WRK):
+#endif
         /* make a stack frame */
         stdu    1,-80(1)
         std     31,72(1)

--- a/coregrind/m_syswrap/syswrap-ppc64-linux.c	2026-04-15 20:59:02.182514914 -0400
+++ b/coregrind/m_syswrap/syswrap-ppc64-linux.c	2026-04-15 21:00:38.520837156 -0400
@@ -77,6 +77,10 @@
 #if defined(VGP_ppc64be_linux)
 "   .align   2\n"
 "   .globl   vgModuleLocal_call_on_new_stack_0_1\n"
+#if _CALL_ELF == 2
+"   .type    vgModuleLocal_call_on_new_stack_0_1,@function\n"
+"vgModuleLocal_call_on_new_stack_0_1:\n"
+#else
 "   .section \".opd\",\"aw\"\n"
 "   .align   3\n"
 "vgModuleLocal_call_on_new_stack_0_1:\n"
@@ -85,6 +89,7 @@
 "   .type    .vgModuleLocal_call_on_new_stack_0_1,@function\n"
 "   .globl   .vgModuleLocal_call_on_new_stack_0_1\n"
 ".vgModuleLocal_call_on_new_stack_0_1:\n"
+#endif
 "   mr    %r1,%r3\n\t"     // stack to %sp
 "   mtlr  %r4\n\t"         // retaddr to %lr
 "   ld 5,0(5)\n\t"         // load f_ptr from f_desc[0]
@@ -212,6 +217,10 @@
 #if defined(VGP_ppc64be_linux)
 "   .align   2\n"
 "   .globl   do_syscall_clone_ppc64_linux\n"
+#if _CALL_ELF == 2
+"   .type    do_syscall_clone_ppc64_linux,@function\n"
+"do_syscall_clone_ppc64_linux:\n"
+#else
 "   .section \".opd\",\"aw\"\n"
 "   .align   3\n"
 "do_syscall_clone_ppc64_linux:\n"
@@ -220,6 +229,7 @@
 "   .type    .do_syscall_clone_ppc64_linux,@function\n"
 "   .globl   .do_syscall_clone_ppc64_linux\n"
 ".do_syscall_clone_ppc64_linux:\n"
+#endif
 "       stdu    1,-64(1)\n"
 "       std     29,40(1)\n"
 "       std     30,48(1)\n"
